cmd/src/doc.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"text/template"
+
+	"github.com/sourcegraph/src-cli/internal/output"
+)
+
+func init() {
+	flagSet := flag.NewFlagSet("doc", flag.ExitOnError)
+	var (
+		outputFlag = flagSet.String("o", "", "Output directory")
+	)
+
+	usage := `
+'src doc' is an internal command that generates the Markdown reference
+documentation used within Sourcegraph.
+
+Usage:
+
+	src doc -o DIR
+	
+Examples:
+
+    $ src doc -o ~/sourcegraph/doc/integration/cli/reference
+	`
+
+	handler := func(args []string) error {
+		flagSet.Parse(args)
+
+		dr, err := newDocRenderer()
+		if err != nil {
+			return err
+		}
+
+		commanders := map[string]*commander{
+			"":             &commands,
+			"campaigns":    &campaignsCommands,
+			"config":       &configCommands,
+			"extensions":   &extensionsCommands,
+			"extsvc":       &extsvcCommands,
+			"lsif":         &lsifCommands,
+			"orgs":         &orgsCommands,
+			"orgs members": &orgsMembersCommands,
+			"repos":        &reposCommands,
+			"users":        &usersCommands,
+		}
+
+		out := output.NewOutput(flagSet.Output(), output.OutputOpts{Verbose: *verbose})
+
+		pending := out.Pending(output.Line("", output.StylePending, "Rendering Markdown..."))
+		count := 0
+		defer func() {
+			pending.Complete(output.Linef(output.EmojiSuccess, output.StyleSuccess, "%d files rendered under %s", count, *outputFlag))
+		}()
+		for groupName, cmdr := range commanders {
+			subcommands := map[string]string{}
+
+			for _, cmd := range *cmdr {
+				name := cmd.flagSet.Name()
+
+				// Figure out the fully qualified name of this command.
+				fqcn := strings.TrimSpace(strings.Join([]string{
+					groupName,
+					name,
+				}, " "))
+
+				if fqcn == "doc" || fqcn == "publish" {
+					continue
+				}
+				pending.Update(fqcn)
+
+				// If this name appears in our commanders map, then this isn't a
+				// real command, and we'll handle it differently.
+				if _, ok := commanders[fqcn]; !ok {
+					content, err := dr.RenderCommand(fqcn, cmd)
+					if err != nil {
+						return err
+					}
+
+					file, err := openDocFile(*outputFlag, fqcn)
+					if err != nil {
+						return err
+					}
+					defer file.Close()
+
+					file.WriteString(content)
+					count++
+					subcommands[name] = name + ".md"
+				} else {
+					subcommands[name] = name + "/index.md"
+				}
+			}
+
+			content, err := dr.RenderGroup(groupName, subcommands)
+			if err != nil {
+				return err
+			}
+
+			file, err := openDocFile(*outputFlag, groupName+" index")
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			file.WriteString(content)
+			count++
+		}
+
+		return nil
+	}
```

Comment:
Do we need to keep track of all the files that were created and then delete those in the output directory we didn't touch?

It's not a blocker, just something to think about, since we might run into the case where old commands leave documentation lying around.

Reply:
I thought about doing that, and could be convinced to add a flag, but I think your cursed `go generate` suggestion in https://github.com/sourcegraph/sourcegraph/pull/15906#discussion_r525911538 might be the right layer to handle that type of cleanup in.